tests/test_wsi_registration.py

```diff
@@ -0,0 +1,34 @@
+import urllib
+
+import numpy as np
+import PIL.Image as Image
+import pytest
+
+from tiatoolbox.tools.registration.wsi_registration import DFBR
+
+
+def test_extract_features():
+    """Test for CNN based feature extraction function."""
+    main_url = "https://tiatoolbox.dcs.warwick.ac.uk/testdata/registration/"
+    urllib.request.urlretrieve(main_url + "HE_1_level8_gray.png", "sample.png")
+    fixed_img = np.asarray(Image.open("sample.png"))
+    urllib.request.urlretrieve(main_url + "HE_2_level8_gray.png", "sample.png")
+    moving_img = np.asarray(Image.open("sample.png"))
+
+    df = DFBR()
+    with pytest.raises(
+        ValueError,
+        match=r".*The required shape for fixed and moving images is n x m x 3.*",
+    ):
+        _ = df.extract_features(fixed_img, moving_img)
+
+    fixed_img = np.repeat(np.expand_dims(fixed_img, axis=2), 1, axis=2)
+    moving_img = np.repeat(np.expand_dims(moving_img, axis=2), 1, axis=2)
+    with pytest.raises(
+        ValueError, match=r".*The input images are expected to have 3 channels.*"
+    ):
+        _ = df.extract_features(fixed_img, moving_img)
+
+    fixed_img = np.repeat(fixed_img, 3, axis=2)
+    moving_img = np.repeat(moving_img, 3, axis=2)
+    _ = df.extract_features(fixed_img, moving_img)
```

Comment:
you should use assert for any checks. e.g., you can use some metrics.

Reply:
This feature extraction step is followed by feature mapping and matching steps. Evaluation metrics will be used at later stages.

Reply:
Please check: https://github.com/TissueImageAnalytics/tiatoolbox/blob/edc7e7f770dc66d154a5ddb26d87b4fed7789d84/tests/test_wsi_registration.py#L46-L55

Reply:
This looks fine but it means the test on line 34 is redundant as it is covered in the code line 46 to 55.

Reply:
The redundant code is removed now.